core/SupportedBrowser.php

```diff
@@ -0,0 +1,69 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik;
+
+use Piwik\Http;
+use Piwik\Piwik;
+use Piwik\Container\StaticContainer;
+use Piwik\DeviceDetector\DeviceDetectorFactory;
+use Piwik\Exception\NotSupportedBrowserException;
+
+class SupportedBrowser
+{
+    /**
+     * A list of browsers with version numbers that are not supported. A browser
+     * is not supported if it's included in this array and the version number
+     * is smaller or equal to the number in this array. If the user's browser
+     * not included in this list or the version number is higher, then
+     * supported.
+     *
+     * Current version numbers are coming from this list:
+     * https://caniuse.com/rel-noreferrer
+     */
+    private static $notSupportedBrowsers = [
+        'FF' => 32,
+        'IE' => 10,
+        'SF' => 4,
+        'CH' => 15,
+        'OP' => 12,
+        'PS' => 12,
+    ];
+
+    public static function checkIfBrowserSupported()
+    {
+        $userAgent = Http::getUserAgent();
+        $ddFactory = StaticContainer::get(DeviceDetectorFactory::class);
+        /** @var \DeviceDetector\DeviceDetector */
+        $deviceDetector = $ddFactory->makeInstance($userAgent);
+
+        $deviceDetector->parse();
+        $client = $deviceDetector->getClient();
+
+        if ($client['type'] === 'browser' && self::browserNotSupported($client['short_name'], (int)$client['version'])) {
```

Comment:
Really random guess (as I don't know PHP well enough):
Is it possible that there is some version string in some minor browser that looks something like `"v3.4.5"`  and will then become `0`?

And what about user agents that are detected as a popular browser, but don't include a version? (maybe instead of freezing the user agent, the browser vendor decides to remove it completely some day in the future to urge people to use feature-detection instead)

UserAgents feel a bit to chaotic for me to decide based on them to completly lock users out of using Matomo

Reply:
@sgiehl probably knows better this, but what I see in the device detector, it tries to get the major version of the client, without any `v` or `.`. Is that correct @sgiehl ?

But maybe a check if the version is 0 would be useful, you are right.

And the other question is the caching. The device detector has some kind of cache, hasn't it @sgiehl ?

I will look into the UI test.

Reply:
I am not entirely sure if that is true for every browser in the detection list, but it might be.

And to answer the question about caching: DeviceDetector does caching, but it caches the conversion from the yaml rule files into one giant regex. The giant regex would still need to be executed every time DD is run, which may or may not take a bit of time.

Reply:
The version should be without any prefix. 
Maybe we could cache the result or at least some kind of compatibility state in the session. Should be unlikely someone switches the useragent within a session 🙈

Reply:
FYI we would be only able to store it in a session if a session was also started ideally. Might not need to add it for now as it might not make it too much slower? It's just that additional caches/sessions make things quite a bit more complicated again and sometimes we wouldn't even start a session maybe in some cases (I think eg when widgetized is executed using token_auth etc) so it's use might be limited. Personally would maybe not add caching by session for now